cluster-autoscaler/utils/nodegroupset/scale_up.go

```diff
@@ -0,0 +1,140 @@
+	NewSize int
+	// MaxSize is the maximum allowed size of the Group
+	MaxSize int
+}
+
+// String is used for printing ScaleUpInfo for logging, etc
+func (s ScaleUpInfo) String() string {
+	return fmt.Sprintf("{%v %v->%v (max: %v)}", s.Group.Id(), s.CurrentSize, s.NewSize, s.MaxSize)
+}
+
+// BalanceScaleUpBetweenGroups distributes a given number of nodes between
+// given set of NodeGroups. The nodes are added to smallest group first, trying
+// to make the group sizes as evenly balanced as possible.
+//
+// Returns ScaleUpInfos for groups that need to be resized.
+//
+// MaxSize of each group will be respected. If newNodes > total free capacity
+// of all NodeGroups it will be capped to total capacity. In particular if all
+// group already have MaxSize, empty list will be returned.
+func BalanceScaleUpBetweenGroups(groups []cloudprovider.NodeGroup, newNodes int) ([]ScaleUpInfo, *errors.AutoscalerError) {
+	if len(groups) == 0 {
+		return []ScaleUpInfo{}, errors.NewAutoscalerError(
+			errors.InternalError, "Can't balance scale up between 0 groups")
+	}
+
+	// get all data from cloudprovider, build data structure
+	scaleUpInfos := make([]ScaleUpInfo, 0)
+	totalCapacity := 0
+	for _, ng := range groups {
+		currentSize, err := ng.TargetSize()
+		if err != nil {
+			return []ScaleUpInfo{}, errors.NewAutoscalerError(
+				errors.CloudProviderError,
+				"failed to get node group size: %v", err)
+		}
+		maxSize := ng.MaxSize()
+		if currentSize == maxSize {
+			// group already maxed, ignore it
+			continue
+		}
+		info := ScaleUpInfo{
+			Group:       ng,
+			CurrentSize: currentSize,
+			NewSize:     currentSize,
+			MaxSize:     maxSize}
+		scaleUpInfos = append(scaleUpInfos, info)
+		totalCapacity += maxSize - currentSize
+	}
+	if totalCapacity < newNodes {
+		glog.Infof("Requested scale-up (%v) exceeds node group set capacity, capping to %v", newNodes, totalCapacity)
+		newNodes = totalCapacity
+	}
+
+	// The actual balancing algorithm.
+	// Sort the node groups by current size and just loop over nodes adding
+	// to smallest group. If a group hits max size remove it from the list
+	// (by moving it to start of the list and increasing startIndex).
+	//
+	// In each iteration we either allocate one node, or 'remove' a maxed out
+	// node group, so this will terminate in O(#nodes + #node groups) steps.
+	// We already know that newNodes <= total capacity, so we don't have to
+	// worry about accidentally removing all node groups while we still
+	// have nodes to allocate.
+	//
+	// Loop invariants:
+	// startIndex <= currentIndex < len(scaleUpInfos)
+	// currentIndex <= i < j -> scaleUpInfos[i] <= scaleUpInfos[j]
+	// startIndex <= i < j < currentIndex -> scaleUpInfos[i] == scaleUpInfos[j]
+	// startIndex <= i < currentIndex <= j -> scaleUpInfos[i] <= scaleUpInfos[j] + 1
+	sort.Slice(scaleUpInfos, func(i, j int) bool {
+		return scaleUpInfos[i].CurrentSize < scaleUpInfos[j].CurrentSize
+	})
+	startIndex := 0
```

Comment:
I have a bunch of unittests that have groups filling up which I think would fail if it didn't work. Also I'm not sure it increases complexity - I have to care about updating it, but I don't have to worry about dealing with groups that already hit MaxSize in the loop. In particular if I don't remove the maxed out groups I lose my invariants related to stuff being sorted by CurrentSize and the condition in `// Update currentIndex` part would have to be more complex to take care of this. 

Less importantly - I think without removing the complexity would be O(#nodes * #node groups) (multiplication instead of addition).